File: js/src/tools/codex.mjs
/**
 * Codex CLI tool configuration
 * Based on hive-mind's codex.lib.mjs implementation
 */

import { buildCommandHead, escapeArg, normalizeExtraArgs } from './shell.mjs';

/**
 * Available Codex model configurations
 * Maps aliases to full model IDs
 */
export const modelMap = {
  gpt5: 'gpt-5',
  'gpt-5': 'gpt-5',
  'gpt5-codex': 'gpt-5-codex',
  // GPT-5.5 family (hive-mind PR #1657, default)
  'gpt-5.5': 'gpt-5.5',
  'gpt-5.5-mini': 'gpt-5.5-mini',
  'gpt-5.5-nano': 'gpt-5.5-nano',
  // GPT-5.4 family
  'gpt-5.4': 'gpt-5.4',
  'gpt-5.4-mini': 'gpt-5.4-mini',
  'gpt-5.4-nano': 'gpt-5.4-nano',
  // GPT-5.3 family (codex variants)
  'gpt-5.3-codex': 'gpt-5.3-codex',
  'gpt-5.3-codex-spark': 'gpt-5.3-codex-spark',
  // GPT-5.2 family
  'gpt-5.2': 'gpt-5.2',
  'gpt-5.2-codex': 'gpt-5.2-codex',
  // GPT-5.1 family
  'gpt-5.1-codex-max': 'gpt-5.1-codex-max',
  o3: 'o3',
  'o3-mini': 'o3-mini',
  gpt4: 'gpt-4',
  'gpt-4': 'gpt-4',
  gpt4o: 'gpt-4o',
  'gpt-4o': 'gpt-4o',
  claude: 'claude-3-5-sonnet',
  sonnet: 'claude-3-5-sonnet',
  opus: 'claude-3-opus',
};

/**
 * Map model alias to full model ID
 * @param {Object} options - Options
 * @param {string} options.model - Model alias or full ID
 * @returns {string} Full model ID
 */
export function mapModelToId(options) {
  const { model } = options;
  return modelMap[model] || model;
}

/**
 * Build command line arguments for Codex
 * @param {Object} options - Options
 * @param {string} [options.prompt] - User prompt (combined with system prompt)
 * @param {string} [options.systemPrompt] - System prompt (prepended to prompt)
 * @param {string} [options.model] - Model to use
 * @param {boolean} [options.json] - JSON output mode
 * @param {string} [options.resume] - Resume session/thread ID
 * @param {boolean} [options.readOnly] - Use Codex read-only sandbox
 * @param {string} [options.sandboxMode] - Explicit Codex sandbox mode
 * @param {string} [options.approvalMode] - Explicit Codex approval mode
 * @param {string[]} [options.extraArgs] - Extra raw CLI args appended after typed args
 * @param {boolean} [options.skipDefaultSafetyFlags] - Do not add default bypass flags
 * @returns {string[]} Array of CLI arguments
 */
export function buildArgs(options) {
  const {
    model,
    json = true,
    resume,
    readOnly = false,
    sandboxMode,
    approvalMode,
    extraArgs = [],
    skipDefaultSafetyFlags = false,
  } = options;

  const args = ['exec'];

  if (resume) {
    args.push('resume', resume);
  }

  if (model) {
    const mappedModel = mapModelToId({ model });
    args.push('--model', mappedModel);
  }

  if (json) {
    args.push('--json');
  }

  // Codex-specific flags for autonomous execution
  args.push('--skip-git-repo-check');
  if (readOnly) {
    args.push('--sandbox', 'read-only');
  } else if (sandboxMode) {
    args.push('--sandbox', sandboxMode);
  }

  if (!readOnly && !sandboxMode && !approvalMode && !skipDefaultSafetyFlags) {
    args.push('--dangerously-bypass-approvals-and-sandbox');
  }

  args.push(...normalizeExtraArgs(extraArgs));

  return args;
}

/**
 * Build complete command string for Codex
 * Codex uses stdin for prompt input
 * @param {Object} options - Options
 * @param {string} options.workingDirectory - Working directory
 * @param {string} [options.prompt] - User prompt
 * @param {string} [options.promptFile] - File containing combined prompt input
 * @param {string} [options.systemPrompt] - System prompt
 * @param {string} [options.model] - Model to use
 * @param {boolean} [options.json] - JSON output mode
 * @param {string} [options.resume] - Resume session ID
 * @param {boolean} [options.readOnly] - Use Codex read-only sandbox
 * @param {string} [options.executable='codex'] - Executable path/name
 * @param {Object|Array} [options.extraEnv] - Environment variables for the tool
 * @param {string[]} [options.extraArgs] - Extra raw CLI args appended after typed args
 * @param {string} [options.sandboxMode] - Explicit Codex sandbox mode
 * @param {string} [options.approvalMode] - Explicit Codex approval mode
 * @param {boolean} [options.skipDefaultSafetyFlags] - Do not add default bypass flags
 * @returns {string} Complete command string
 */
export function buildCommand(options) {
  const {
    prompt,
    promptFile,
    systemPrompt,
    executable = 'codex',
    extraEnv,
    ...argOptions
  } = options;
  const args = buildArgs(argOptions);

  // Codex expects prompt via stdin, combine system and user prompts
  const combinedPrompt = systemPrompt
    ? `${systemPrompt}\n\n${prompt || ''}`
    : prompt || '';

  // Build command with stdin piping
  const inputCommand = promptFile
    ? `cat ${escapeArg(promptFile)}`
    : `printf '%s' '${combinedPrompt.replace(/'/g, "'\\''")}'`;
  const prefixArgs = [];
  if (argOptions.readOnly) {
    prefixArgs.push('--ask-for-approval', 'never');
  } else if (argOptions.approvalMode) {
    prefixArgs.push('--ask-for-approval', argOptions.approvalMode);
  }

  return `${inputCommand} | ${buildCommandHead({
    executable,
    extraEnv,
    prefixArgs,
  })} ${args.map(escapeArg).join(' ')}`.trim();
}

/**
 * Parse JSON messages from Codex output
 * Codex outputs NDJSON format
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {Object[]} Array of parsed JSON messages
 */
export function parseOutput(options) {
  const { output } = options;
  const messages = [];
  const lines = output.split('\n');

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || !trimmed.startsWith('{')) {
      continue;
    }

    try {
      const parsed = JSON.parse(trimmed);
      messages.push(parsed);
    } catch {
      // Skip lines that aren't valid JSON
    }
  }

  return messages;
}

/**
 * Extract session/thread ID from Codex output
 * Codex uses thread_id instead of session_id
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {string|null} Session ID or null
 */
export function extractSessionId(options) {
  const { output } = options;
  const messages = parseOutput({ output });

  for (const msg of messages) {
    if (msg.thread_id) {
      return msg.thread_id;
    }
    if (msg.session_id) {
      return msg.session_id;
    }
  }

  return null;
}

/**
 * Extract usage statistics from Codex output
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {Object} Usage statistics
 */
export function extractUsage(options) {
  const { output } = options;
  const messages = parseOutput({ output });

  const usage = {
    inputTokens: 0,
    outputTokens: 0,
  };

  for (const msg of messages) {
    if (msg.usage) {
      const u = msg.usage;
      if (u.input_tokens) {
        usage.inputTokens += u.input_tokens;
      }
      if (u.output_tokens) {
        usage.outputTokens += u.output_tokens;
      }
    }
  }

  return usage;
}

/**
 * Codex tool configuration
 */
export const codexTool = {
  name: 'codex',
  displayName: 'Codex CLI',
  executable: 'codex',
  supportsJsonOutput: true,
  supportsJsonInput: true, // Codex can accept JSON input via stdin
  supportsSystemPrompt: false, // System prompt is combined with user prompt
  supportsResume: true,
  supportsReadOnly: true, // Supports --sandbox read-only
  defaultModel: 'gpt-5.5', // hive-mind PR #1657
  modelMap,
  mapModelToId,
  buildArgs,
  buildCommand,
  parseOutput,
  extractSessionId,
  extractUsage,
};

File: rust/src/tools/codex.rs
//! Codex CLI tool configuration
//! Based on hive-mind's codex.lib.mjs implementation

use crate::streaming::parse_ndjson;
use crate::tools::shell::{build_command_head, escape_arg, escape_single_quotes};
use serde_json::Value;
use std::collections::HashMap;

/// Get the Codex model map
pub fn get_model_map() -> HashMap<&'static str, &'static str> {
    let mut map = HashMap::new();
    map.insert("gpt5", "gpt-5");
    map.insert("gpt-5", "gpt-5");
    map.insert("gpt5-codex", "gpt-5-codex");
    // GPT-5.5 family (hive-mind PR #1657, default)
    map.insert("gpt-5.5", "gpt-5.5");
    map.insert("gpt-5.5-mini", "gpt-5.5-mini");
    map.insert("gpt-5.5-nano", "gpt-5.5-nano");
    // GPT-5.4 family
    map.insert("gpt-5.4", "gpt-5.4");
    map.insert("gpt-5.4-mini", "gpt-5.4-mini");
    map.insert("gpt-5.4-nano", "gpt-5.4-nano");
    // GPT-5.3 family (codex variants)
    map.insert("gpt-5.3-codex", "gpt-5.3-codex");
    map.insert("gpt-5.3-codex-spark", "gpt-5.3-codex-spark");
    // GPT-5.2 family
    map.insert("gpt-5.2", "gpt-5.2");
    map.insert("gpt-5.2-codex", "gpt-5.2-codex");
    // GPT-5.1 family
    map.insert("gpt-5.1-codex-max", "gpt-5.1-codex-max");
    map.insert("o3", "o3");
    map.insert("o3-mini", "o3-mini");
    map.insert("gpt4", "gpt-4");
    map.insert("gpt-4", "gpt-4");
    map.insert("gpt4o", "gpt-4o");
    map.insert("gpt-4o", "gpt-4o");
    map.insert("claude", "claude-3-5-sonnet");
    map.insert("sonnet", "claude-3-5-sonnet");
    map.insert("opus", "claude-3-opus");
    map
}

/// Map model alias to full model ID
///
/// # Arguments
/// * `model` - Model alias or full ID
///
/// # Returns
/// Full model ID
pub fn map_model_to_id(model: &str) -> String {
    let model_map = get_model_map();
    model_map
        .get(model)
        .map(|s| s.to_string())
        .unwrap_or_else(|| model.to_string())
}

/// Codex command build options
#[derive(Debug, Clone, Default)]
pub struct CodexBuildOptions {
    pub prompt: Option<String>,
    pub prompt_file: Option<String>,
    pub system_prompt: Option<String>,
    pub model: Option<String>,
    pub json: bool,
    pub resume: Option<String>,
    pub read_only: bool,
    pub executable: Option<String>,
    pub extra_env: Vec<(String, String)>,
    pub extra_args: Vec<String>,
    pub skip_default_safety_flags: bool,
    pub sandbox_mode: Option<String>,
    pub approval_mode: Option<String>,
}

/// Build command line arguments for Codex
///
/// # Arguments
/// * `options` - Build options
///
/// # Returns
/// Vector of CLI arguments
pub fn build_args(options: &CodexBuildOptions) -> Vec<String> {
    let mut args = vec!["exec".to_string()];

    if let Some(ref resume) = options.resume {
        args.push("resume".to_string());
        args.push(resume.clone());
    }

    if let Some(ref model) = options.model {
        let mapped_model = map_model_to_id(model);
        args.push("--model".to_string());
        args.push(mapped_model);
    }

    // Default to json=true like JavaScript version
    if options.json {
        args.push("--json".to_string());
    }

    // Codex-specific flags for autonomous execution
    args.push("--skip-git-repo-check".to_string());
    if options.read_only {
        args.push("--sandbox".to_string());
        args.push("read-only".to_string());
    } else if let Some(ref sandbox_mode) = options.sandbox_mode {
        args.push("--sandbox".to_string());
        args.push(sandbox_mode.clone());
    }

    if !options.read_only
        && options.sandbox_mode.is_none()
        && options.approval_mode.is_none()
        && !options.skip_default_safety_flags
    {
        args.push("--dangerously-bypass-approvals-and-sandbox".to_string());
    }

    args.extend(options.extra_args.clone());

    args
}

/// Build complete command string for Codex
/// Codex uses stdin for prompt input
///
/// # Arguments
/// * `options` - Build options
///
/// # Returns
/// Complete command string
pub fn build_command(options: &CodexBuildOptions) -> String {
    let args = build_args(options);
    let args_str: Vec<String> = args.iter().map(|a| escape_arg(a)).collect();

    // Codex expects prompt via stdin, combine system and user prompts
    let combined_prompt = match (&options.system_prompt, &options.prompt) {
        (Some(sys), Some(prompt)) => format!("{}\n\n{}", sys, prompt),
        (Some(sys), None) => sys.clone(),
        (None, Some(prompt)) => prompt.clone(),
        (None, None) => String::new(),
    };

    // Build command with stdin piping
    let input_command = options.prompt_file.as_ref().map_or_else(
        || format!("printf '%s' '{}'", escape_single_quotes(&combined_prompt)),
        |prompt_file| format!("cat {}", escape_arg(prompt_file)),
    );
    let executable = options.executable.as_deref().unwrap_or("codex");
    let mut prefix_args = Vec::new();
    if options.read_only {
        prefix_args.push("--ask-for-approval".to_string());
        prefix_args.push("never".to_string());
    } else if let Some(ref approval_mode) = options.approval_mode {
        prefix_args.push("--ask-for-approval".to_string());
        prefix_args.push(approval_mode.clone());
    }

    format!(
        "{} | {} {}",
        input_command,
        build_command_head(executable, &options.extra_env, &prefix_args),
        args_str.join(" ")
    )
    .trim()
    .to_string()
}

/// Parse JSON messages from Codex output
/// Codex outputs NDJSON format
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Vector of parsed JSON messages
pub fn parse_output(output: &str) -> Vec<Value> {
    parse_ndjson(output)
}

/// Extract session/thread ID from Codex output
/// Codex uses thread_id instead of session_id
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Session ID or None
pub fn extract_session_id(output: &str) -> Option<String> {
    let messages = parse_output(output);

    for msg in messages {
        if let Some(thread_id) = msg.get("thread_id").and_then(|v| v.as_str()) {
            return Some(thread_id.to_string());
        }
        if let Some(session_id) = msg.get("session_id").and_then(|v| v.as_str()) {
            return Some(session_id.to_string());
        }
    }

    None
}

/// Usage statistics
#[derive(Debug, Clone, Default)]
pub struct CodexUsage {
    pub input_tokens: u64,
    pub output_tokens: u64,
}

/// Extract usage statistics from Codex output
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Usage statistics
pub fn extract_usage(output: &str) -> CodexUsage {
    let messages = parse_output(output);
    let mut usage = CodexUsage::default();

    for msg in messages {
        if let Some(msg_usage) = msg.get("usage") {
            if let Some(input) = msg_usage.get("input_tokens").and_then(|v| v.as_u64()) {
                usage.input_tokens += input;
            }
            if let Some(output) = msg_usage.get("output_tokens").and_then(|v| v.as_u64()) {
                usage.output_tokens += output;
            }
        }
    }

    usage
}

/// Codex tool configuration
#[derive(Debug, Clone)]
pub struct CodexTool {
    pub name: &'static str,
    pub display_name: &'static str,
    pub executable: &'static str,
    pub supports_json_output: bool,
    pub supports_json_input: bool,
    pub supports_system_prompt: bool,
    pub supports_resume: bool,
    pub supports_read_only: bool,
    pub default_model: &'static str,
}

impl Default for CodexTool {
    fn default() -> Self {
        Self {
            name: "codex",
            display_name: "Codex CLI",
            executable: "codex",
            supports_json_output: true,
            supports_json_input: true, // Codex can accept JSON input via stdin
            supports_system_prompt: false, // System prompt is combined with user prompt
            supports_resume: true,
            supports_read_only: true, // Supports --sandbox read-only
            default_model: "gpt-5.5", // hive-mind PR #1657
        }
    }
}

// Tests are in rust/tests/codex_tests.rs
